llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -629,7 +629,10 @@ RISCVTargetLowering::RISCVTargetLowering(const TargetMachine &TM,
 
   if (Subtarget.hasStdExtA()) {
     setMaxAtomicSizeInBitsSupported(Subtarget.getXLen());
-    setMinCmpXchgSizeInBits(32);
+    if (Subtarget.hasStdExtZabha())
+      setMinCmpXchgSizeInBits(8);
```

Comment:
Is this accurate? We would need LR/SC for bytes and half words, or Zacas to be able to cmpxchg on 8 bits.

Reply:
I see now why you had to change this since it's used to by the AtomicExpand pass to know if masking is needed. But changing it produces incorrect code for 8 and 16 bit Nand and cmpxchg where we still need masking.

Reply:
Without this change, I cannot emit `amoxor.b/amoand.b/amoor.b` since AtomicExpand pass always widen them to 32-bit:

https://github.com/llvm/llvm-project/blob/995d21bc6ff2220b2887cf9640d936eb99b3c617/llvm/lib/CodeGen/AtomicExpandPass.cpp#L325-L333

Do you have any idea on how to fix it?

Reply:
That came from https://github.com/llvm/llvm-project/commit/3291f9aa8168408ce5ef4757012c06c196a72c41, which seems slightly odd. Seems to be assuming shouldExpandAtomicRMWInIR doesn't return None, which it probably should be checking?

Reply:
> That came from [3291f9a](https://github.com/llvm/llvm-project/commit/3291f9aa8168408ce5ef4757012c06c196a72c41), which seems slightly odd. Seems to be assuming shouldExpandAtomicRMWInIR doesn't return None, which it probably should be checking?

@asb @jyknight Any thoughts?

Reply:
ctopper fixed in #80947, so this should be fine to change to `Subtarget.hasStdExtZabha && Subtarget.hasStdExtZacas()` now.